### PR TITLE
Peer-based HTTP sync for new blocks, chain data and pool

### PR DESCRIPTION
**Pull request by damirka**
What we have here:

- good to go block producer which emits 1 block every 5 seconds
- transport infrastructure that involves any client connected to network
- temporary HTTP server for pulling/peering any data

How to run&test:

- start bp (with or without existing chain)
- start index any number of times (each one will pull historical and new data from bp)

What's good:

- new block is shared with all connected clients
- block receiving starts only after client has been synced
- sync process is set up
- nice logging of all tech info for debugging

What's not so good:

- only one block producer possible (bp choosing logic remains uncompleted)
- there may be lag when chain data is too big and one block is missed (still hypothetical)
- chain and pool data files are temporarily random (you can force them via env)
- I get odd feeing about all of this (not sure why)

### CODE REVIEW

File: docs/SYNC_MECHANICS.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Pool/Chain sync mechanics
+
+To provide fast setup of new node in this PoC application while keeping decentralisation.
+
+1. New node emits event **REQUEST_CHAIN**
+2. All the nodes in network receive it and emit number of blocks (or txs count in pool case) using **SHARE_CHAIN** event
+3. Newbie decides from which node to pull data and sends **CREATE_CHAINDATA_SERVER**
+4. Node with info sets up one-time HTTP server and sends **CHAINDATA_SERVER_CREATED** directly to newbit
```
Comment (romfrolov) [inline docs/SYNC_MECHANICS.md]:
```suggestion
4. Node with info sets up one-time HTTP server and sends **CHAINDATA_SERVER_CREATED** directly to newbie
```

File: network/waiter.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+const msg = require('core/transport');
+
+exports.waitFor    = waitFor;
+exports.waitForAll = waitForAll;
+
+/**
+ * Wait for N number of emitted events or for K milliseconds.
+ * This method helps collecting all emitted events in application
+ *
+ * @param  {String}  evt         Name of the event ro await in transport
```
Comment (romfrolov) [inline network/waiter.js]:
```suggestion
 * @param  {String}  evt         Name of the event to await in transport
```

File: index.js [hunk 3]
```diff
@@ -8,9 +8,92 @@
...
+    }
+
+})().then(function runClient() {
+
+    console.log('Starting observer');
+
+    require('network/observer');
+
+    tp.on(events.NEW_BLOCK, function newBlock({block}, msg) {
+        console.log('Received block: %s', JSON.stringify(block));
```
Comment (romfrolov) [inline index.js:47]: The output will be long.

File: index.js [hunk 4]
```diff
@@ -8,9 +8,92 @@
...
+        console.log('Message:');
+        console.log(msg);
+    });
+
+});
+
```
Comment (romfrolov) [inline index.js:53]:
```suggestion

/**
 * Event request timeout.
 *
 * @type {Number}
 */
```

File: core/transport.js [hunk 5]
```diff
@@ -43,108 +49,97 @@ const TARGET_ALL = '*';
...
- */
-const identifiableLayer = (function initIdentifiableLayer() {
+udp.on('message', function (rawData, meta) {
 
-    const groups     = [];
-    const knownNodes = new Map();
-    const emitter    = new Emitter();
+    let msg;
 
-    emitter.joinGroup = function joinGroup(name) {
```
Comment (romfrolov) [inline core/transport.js:93]: How do you join the group now?

**Review by romfrolov — APPROVED** [review]

File: index.js (hunk 3 above)
Comment (damirka) [reply index.js:47]: Nope. There's no actual block in UDP transport. 
There's only shortened version: hash, Phash and number.

File: core/transport.js (hunk 5 above)
Comment (damirka) [reply core/transport.js:93]: Now transport.groups is a Set. 
```
transport.groups.add('name');
transport.groups.delete('name');
```

**Pull request merged**